src/Xamarin.Android.Build.Tasks/Utilities/XAJavaTypeScanner.cs

```diff
@@ -0,0 +1,167 @@
+	{
+		var types = new Dictionary<string, TypeData> (StringComparer.Ordinal);
+		var stopwatch = new Stopwatch ();
+		foreach (ITaskItem asmItem in inputAssemblies) {
+			AndroidTargetArch arch = GetTargetArch (asmItem);
+
+			stopwatch.Start ();
+			AssemblyDefinition asmdef = resolver.GetAssembly (asmItem.ItemSpec);
+			stopwatch.Stop ();
+			log.LogMessage ($"Load of assembly '{asmItem.ItemSpec}', elapsed: {stopwatch.Elapsed}");
+			stopwatch.Reset ();
+
+			stopwatch.Start ();
+			foreach (ModuleDefinition md in asmdef.Modules) {
+				foreach (TypeDefinition td in md.Types) {
+					AddJavaType (td, types, arch);
+				}
+			}
+			stopwatch.Stop ();
+			log.LogMessage ($"Add all types from assembly '{asmItem.ItemSpec}', elapsed: {stopwatch.Elapsed}");
+			stopwatch.Reset ();
+		}
+
+		var ret = new List<JavaType> ();
+		foreach (var kvp in types) {
+			ret.Add (new JavaType (kvp.Value.FirstType, kvp.Value.IsAbiSpecific ? kvp.Value.PerAbi : null));
+		}
+
+		return ret;
+	}
+
+	void AddJavaType (TypeDefinition type, Dictionary<string, TypeData> types, AndroidTargetArch arch)
+	{
+		if (type.IsSubclassOf ("Java.Lang.Object", cache) || type.IsSubclassOf ("Java.Lang.Throwable", cache) || (type.IsInterface && type.ImplementsInterface ("Java.Interop.IJavaPeerable", cache))) {
+			// For subclasses of e.g. Android.App.Activity.
+			string typeName = type.GetPartialAssemblyQualifiedName (cache);
+			if (!types.TryGetValue (typeName, out TypeData typeData)) {
+				typeData = new TypeData (type);
+				types.Add (typeName, typeData);
+			}
+
+			if (typeData.PerAbi.ContainsKey (AndroidTargetArch.None)) {
+				if (arch == AndroidTargetArch.None) {
+					throw new InvalidOperationException ($"Duplicate type '{type.FullName}' in assembly {type.Module.FileName}");
+				}
+
+				throw new InvalidOperationException ($"Previously added type '{type.FullName}' was in ABI-agnostic assembly, new one comes from ABI {arch} assembly");
+			}
+
+			if (typeData.PerAbi.ContainsKey (arch)) {
+				throw new InvalidOperationException ($"Duplicate type '{type.FullName}' in assembly {type.Module.FileName}, for ABI {arch}");
+			}
+
+			typeData.PerAbi.Add (arch, type);
+		} else if (type.IsClass && !type.IsSubclassOf ("System.Exception", cache) && type.ImplementsInterface ("Android.Runtime.IJavaObject", cache)) {
+			string message = $"XA4212: Type `{type.FullName}` implements `Android.Runtime.IJavaObject` but does not inherit `Java.Lang.Object` or `Java.Lang.Throwable`. This is not supported.";
+
+			if (ErrorOnCustomJavaObject) {
+				log.LogError (message);
+			} else {
+				log.LogWarning (message);
+			}
+			return;
+		}
+
+		if (!type.HasNestedTypes) {
+			return;
+		}
+
+		foreach (TypeDefinition nested in type.NestedTypes) {
+			AddJavaType (nested, types, arch);
+		}
+	}
+
+	AndroidTargetArch GetTargetArch (ITaskItem asmItem)
+	{
+		string? abi = asmItem.GetMetadata ("Abi");
```

Comment:
This should probably "sniff the path" in addition to using the `%(Abi)` metadata?

Reply:
No, `Abi` is added based on the path elsewhere in our targets. The metadata was added specifically for this purpose sometime ago.